docs: Add recipe for using getImage() (#5042)
withastro/docs: src/content/docs/en/recipes/build-custom-img-component.mdx

---
title: Build a custom image component 
description: Learn how to build a custom image component that supports media queries using the getImage function
i18nReady: true
type: recipe
---

> Astro provides a couple of built in components that you can use to display and optimize your images.  The `<Picture>` component allows you to display responsive images and work with different formats and sizes. The `<Image>` componet will optimze your images and allow you to pass in different formats and quality properties.

Reviewer: ```suggestion
Astro provides a two built-in components that you can use to display and optimize your images.  The `<Picture>` component allows you to display responsive images and work with different formats and sizes. The `<Image>` component will optimize your images and allow you to pass in different formats and quality properties.
```

But there will be times where you will need to have more customizable options for your images that the `<Picture>` and `<Image>` components currently do not support. 

> In this recipe, you will learn how use the `getImage()` function to create your own custom image component that supports the use of media queries/art direction.  The purpose of the `getImage()` function is to create images for use outside of direct embedding in HTML.

Reviewer: ```suggestion
In this recipe, you will learn how use the [`getImage()` function](/en/guides/images/#generating-images-with-getimage) to create your own custom image component that supports the use of media queries/art direction.
```
I think I like this better as a shorter statement, with a link to `getImage()`!

## Recipe

1. Create a new Astro component and import the `getImage()` function

    ```astro title="src/components/MyCustomImageComponent.astro"
    ---
     import { getImage } from "astro:assets";
    ---

    ```

> 2. Destructure the `props` from `Astro.props`. You can also use TypeScript to type the props. 

Reviewer: ```suggestion
2. Destructure three `props` from `Astro.props`: `mobileImgUrl` and `desktopImgUrl` for creating your image at different viewport sizes, as well as `alt` for the image's alt text. These props will be passed wherever you render your custom image components. You can also use TypeScript to type the props. 
```
So, what about something kind of like this? It felt a little early to say "destructure THE props" when we haven't seen anything yet? So, I'm thinking of something that introduces them a tiny bit?

Also, while destructure is totally fine, you could also consider a phrasing something like:

> Your custom component will receive/use three props..
or
> Create three props that your custom component will receive...

Again, nothing wrong with using destructure, but if the emphasis is now on introducing these props that will be created, then maybe it's not as easy to squish those ideas into a sentence starting with destructure. So, try some things, and see what you like!

Author: yeah, the notes about destructuring makes sense here.
I will make some updates 👍

>    ```astro title="src/components/MyCustomImageComponent.astro"

Reviewer: ```suggestion
   ```astro title="src/components/MyCustomImageComponent.astro" ins={11}
```
Up to you! Just making sure you're aware of our syntax highlighting options that we can show inserted lines as green by marking the line numbers like this. If there are only certain words or phrases you want to show inserted, you can do strings too, like `ins="getImage"`

Author: oh that is really cool. I will go ahead and add that in 👍

    ---
    import type { ImageMetadata } from "astro";
    import { getImage } from "astro:assets";

    interface Props {
        mobileImgUrl: string | ImageMetadata;
        desktopImgUrl: string | ImageMetadata;
        alt: string;
    }

    const { mobileImgUrl, desktopImgUrl, alt } = Astro.props;
    ---

    ```

> 3. Call the `getImage()` function and pass in the properties you want to use into the options object 

Reviewer: ```suggestion
3. Call the `getImage()` function and pass in the properties you want to use into the options object to define each separate image you want to create.
```
Something like this maybe? To emphasize we're doing this for each responsive image size?

I could also see something phrased in the opposite direction, like:

> Define each of your responsive images by calling the `getImage()` function with your desired properties.

Whatever you like!

Author: I like the second option 👍

    ```astro title="src/components/MyCustomImageComponent.astro"
    ---
    import type { ImageMetadata } from "astro";
    import { getImage } from "astro:assets";

    interface Props {
        mobileImgUrl: string | ImageMetadata;
        desktopImgUrl: string | ImageMetadata;
        alt: string;
    }

    const { mobileImgUrl, desktopImgUrl, alt } = Astro.props;

    const mobileImg = await getImage({
        src: mobileImgUrl,
        format: "webp",
        width: 200,
        height: 200,
    });

    const desktopImg = await getImage({
        src: desktopImgUrl,
        format: "webp",
        width: 800,
        height: 200,
    });
    ---

    ```

> 4. Add the `source` and `img` elements inside the `picture` element and then set your desired media queries. 

Reviewer: ```suggestion
4. Create a `<picture>` element that generates `srcset` with your different images based on your desired media queries. 
```
If the picture element is a new thing, I always like instructions that say, K, you gotta make a picture element with.... I know these are tiny word changes, but when people read "THE" thing, it can stress them out if they don't know exactly what that is. ("THE picture element? Wait, WHAT picture element? I don't know what that is?) Sometimes even just choosing to write "a" instead of "the" feels less like, "I'm supposed to know this exact thing, and I don't."

   ```astro title="src/components/MyCustomImageComponent.astro"
    ---
    import type { ImageMetadata } from "astro";
    import { getImage } from "astro:assets";

    interface Props {
        mobileImgUrl: string | ImageMetadata;
        desktopImgUrl: string | ImageMetadata;
        alt: string;
    }

    const { mobileImgUrl, desktopImgUrl, alt } = Astro.props;

    const mobileImg = await getImage({
        src: mobileImgUrl,
        format: "webp",
        width: 200,
        height: 200,
    });

    const desktopImg = await getImage({
        src: desktopImgUrl,
        format: "webp",
        width: 800,
        height: 200,
    });
    ---

    <picture>
        <source media="(max-width: 799px)" srcset={mobileImg.src} />
        <source media="(min-width: 800px)" srcset={desktopImg.src} />
        <img src={desktopImg.src} alt={alt} />
    </picture>

>     ```

Reviewer: ``````suggestion
    ```

5. Import and use `<MyCustomImageComponent />` in any `.astro` file. Be sure to pass the necessary props for generating two different images at the different viewport sizes:

	```astro title=""
	---
	import .....
	---
	`<MyCustomImageComponent mobileImgUrl="..." desktopImgUrl="..." alt="..." />`
``````
I'd love to see a follow up step that actually uses this component in another Astro component or page! Make up some scenario and show this being used to really finish off this recipe!

Author: I went ahead and added in the new step 5 👍

    import { getImage } from "astro:assets";

    interface Props {
        mobileImgUrl: string | ImageMetadata;
        desktopImgUrl: string | ImageMetadata;
        alt: string;
    }

    const { mobileImgUrl, desktopImgUrl, alt } = Astro.props;
    ---

    ```

3. Define each of your responsive images by calling the `getImage()` function with your desired properties.

    ```astro title="src/components/MyCustomImageComponent.astro" ins={13-18, 20-25}
    ---
    import type { ImageMetadata } from "astro";
    import { getImage } from "astro:assets";

    interface Props {
        mobileImgUrl: string | ImageMetadata;
        desktopImgUrl: string | ImageMetadata;
        alt: string;
    }

    const { mobileImgUrl, desktopImgUrl, alt } = Astro.props;

    const mobileImg = await getImage({
        src: mobileImgUrl,
        format: "webp",
        width: 200,
        height: 200,
    });

    const desktopImg = await getImage({
        src: desktopImgUrl,
        format: "webp",
        width: 800,
        height: 200,
    });
    ---

    ```

4. Create a `<picture>` element that generates a `srcset` with your different images based on your desired media queries. 

>    ```astro title="src/components/MyCustomImageComponent.astro" ins={28-32}

Reviewer 2: 看一下语法高亮怎么丢了
